smart_open/s3.py

```diff
@@ -279,47 +279,98 @@ class _SeekableRawReader(object):
         self._body = None
-        self._position = position
 
-    def _load_body(self):
-        """Build a continuous connection with the remote peer starts from the current postion.
-        """
-        range_string = smart_open.utils.make_range_string(self._position)
-        logger.debug('content_length: %r range_string: %r', self._content_length, range_string)
+        start = None
+        stop = None
+        if whence == constants.WHENCE_START:
+            start = max(0, offset)
+        elif whence == constants.WHENCE_CURRENT:
+            start = max(0, offset + self._position)
+        else:
+            stop = max(0, -offset)
 
-        if self._position == self._content_length == 0 or self._position == self._content_length:
-            #
-            # When reading, we can't seek to the first byte of an empty file.
-            # Similarly, we can't seek past the last byte.  Do nothing here.
-            #
+        if self._content_length is not None and \
+                ((start is not None and start >= self._content_length) or stop == 0):
+            # avoid calling _load_body() if we are at the end of the file
             self._body = io.BytesIO()
+            self._position = self._content_length
         else:
-            self._body = _get(
+            range_string = smart_open.utils.make_range_string(start, stop)
+            self._load_body(range_string)  # sets self._position
+
+        return self._position
+
+    def _load_body(self, range_string=None):
+        """Build a continuous connection with the remote peer starts from the current postion.
+        """
+        if range_string is None:
+            range_string = smart_open.utils.make_range_string(self._position)
+        logger.debug('range_string: %r', range_string)
+
+        try:
+            # Optimistically try to fetch the requested content range.
+            response = _get(
                 self._object,
                 version=self._version_id,
                 Range=range_string,
                 **self._object_kwargs
-            )['Body']
+            )
+        except IOError as error:
+            # If we didn't know _content_length yet in .seek(), we may have requested an
+            # invalid range. In that case, set _position and _content_length from the actual
+            # object size returned by S3.
+            orig_error = error.__context__
+            aws_error = (
+                orig_error
+                and hasattr(orig_error, 'response')
+                and orig_error.response.get('Error', {})
+            )
+            if aws_error and aws_error.get('Message') == 'Requested Range Not Satisfiable':
+                if 'ActualObjectSize' in aws_error:
+                    self._position = self._content_length = int(aws_error['ActualObjectSize'])
+                else:
+                    # this shouldn't happen with real S3,
+                    # but does happen with moto which seems to lack ActualObjectSize
```

Comment:
Are you sure this is the case?

If yes, then please create an issue in the moto repository, describe the problem there, and link to the new issue in your code. We can deal with moto's quirks, but we don't want them in our code indefinitely, so if we include this in our code base, we want to connect it to a future resolution.

Reply:
Filed issue and linked.